### ISSUE TITLE
Flutter error in integration testing

### ISSUE THREAD

**Issue by hanseltertiusJCU** [flutter/flutter#33921]
Hi everyone, I recently got an issue regarding to this, when I tried to do the widget testing, this is the message that I have been getting and I haven't get the solution related to this.

Steps to reproduce:
**home.dart :**
```
import 'package:flutter/material.dart';
import 'package:meals_catalogue/const_strings.dart';
import 'package:meals_catalogue/key_tooltip_strings.dart';
import 'package:meals_catalogue/widgets/data_widget.dart';
 
appBar: AppBar(
  key: Key(getStringKey(APP_BAR)),
  title: appBarTitle,
  actions: getMenuIcon(_dataWidgetsList[_currentIndex].searchEnabled),
  textTheme: TextTheme(
    title: TextStyle(
        fontSize: 20.0,
        fontWeight: FontWeight.bold,
        fontFamily: 'Nunito'),
  ),
),
```
**finder.dart**
```
import 'package:flutter_driver/flutter_driver.dart';
import 'package:meals_catalogue/const_strings.dart';
import 'package:meals_catalogue/key_tooltip_strings.dart';
 
SerializableFinder appBar = find.byValueKey(getStringKey(APP_BAR));
key_tooltip_strings.dart

String getStringKey(String key) {
  return key;
}
 
String getStringTooltip(String tooltip){
  return tooltip;
}
```
**const_strings.dart**
```
// App bar section
const String APP_BAR = "appBar";
 
// Navigation section
const String PAGE_VIEW = "pageView";
const String BOTTOM_NAVIGATION_BAR = "bottomNavigationBar";
 
// Menu in home section
const String TEXT_FIELD = "textField";
const String MENU_ICON_BUTTON = "menuIconButton";
 
// Navigation part section
const String DESSERT = 'dessert';
const String SEAFOOD = 'seafood';
const String FAVORITE_DESSERT = "favoriteDessert";
const String FAVORITE_SEAFOOD = "favoriteSeafood";
 
// Inkwell section
const String INKWELL = "inkWell";
```
**integration test file:**
```
void main(){
  // todo: verify the behavior
  group('Meals Catalogue App', (){
    
    // home
 
    FlutterDriver flutterDriver;
 
    setUpAll(() async {
      flutterDriver = await FlutterDriver.connect();
    });
 
    // todo: app bar title
    test('Show default app bar title', () async {
      // todo: text app bar title
      expect(await flutterDriver.getText(appBar), "Dessert");
    });
    
    tearDownAll(() async {
      if(flutterDriver != null){
        flutterDriver.close();
      }
    });
  });
}
```

**Code error:**
```
D:\AndroidStudioProjects\MFDEProject\meals_catalogue>flutter` drive --target=test_driver/integration_ap
p.dart --flavor develop
Using device Android SDK built for x86.
Starting application: test_driver/integration_app.dart
Initializing gradle...                                              3.8s
Resolving dependencies...                                          51.7s
Installing build\app\outputs\apk\app.apk...                         3.5s
Running Gradle task 'assembleDevelopDebug'...
Running Gradle task 'assembleDevelopDebug'... Done                 52.6s
Built build\app\outputs\apk\develop\debug\app-develop-debug.apk.
Installing build\app\outputs\apk\app.apk...                         3.1s
I/flutter ( 6250): Observatory listening on http://127.0.0.1:35772/s-KXXLpqDeQ=/
00:00 +0: Meals Catalogue App (setUpAll)
[info ] FlutterDriver: Connecting to Flutter application at http://127.0.0.1:57523/s-KXXLpqDeQ=/
[trace] FlutterDriver: Isolate found with number: 355225911
[trace] FlutterDriver: Isolate is paused at start.
[trace] FlutterDriver: Attempting to resume isolate
[trace] FlutterDriver: Waiting for service extension
[info ] FlutterDriver: Connected to Flutter application.
00:02 +0: Meals Catalogue App Show default app bar title
[warning] FlutterDriver: get_text message is taking a long time to complete...
00:32 +0 -1: Meals Catalogue App Show default app bar title [E]
  TimeoutException after 0:00:30.000000: Test timed out after 30 seconds.
  package:test_api/src/backend/invoker.dart 332:28              Invoker._handleError.<fn>
  dart:async/zone.dart 1120:38                                  _rootRun
  dart:async/zone.dart 1021:19                                  _CustomZone.run
  package:test_api/src/backend/invoker.dart 330:10              Invoker._handleError
  package:test_api/src/backend/invoker.dart 283:9               Invoker.heartbeat.<fn>.<fn>
  dart:async/zone.dart 1124:13                                  _rootRun
  dart:async/zone.dart 1021:19                                  _CustomZone.run
  package:test_api/src/backend/invoker.dart 281:38              Invoker.heartbeat.<fn>
  package:stack_trace/src/stack_zone_specification.dart 209:15  StackZoneSpecification._run
  package:stack_trace/src/stack_zone_specification.dart 119:48  StackZoneSpecification._registerCallba
ck.<fn>
  dart:async/zone.dart 1124:13                                  _rootRun
  dart:async/zone.dart 1021:19                                  _CustomZone.run
  dart:async/zone.dart 947:23                                   _CustomZone.bindCallback.<fn>
  dart:async-patch/timer_patch.dart 21:15                       Timer._createTimer.<fn>
  dart:isolate-patch/timer_impl.dart 382:19                     _Timer._runTimers
  dart:isolate-patch/timer_impl.dart 416:5                      _Timer._handleMessage
  dart:isolate-patch/isolate_patch.dart 171:12                  _RawReceivePortImpl._handleMessage
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1045:19                                  _CustomZone.registerCallback
  dart:async/zone.dart 946:22                                   _CustomZone.bindCallback
  dart:async/zone.dart 1191:21                                  _rootCreateTimer
  dart:async/zone.dart 1088:19                                  _CustomZone.createTimer
  package:test_api/src/backend/invoker.dart 280:34              Invoker.heartbeat
  package:test_api/src/backend/invoker.dart 243:5               Invoker.waitForOutstandingCallbacks
  package:test_api/src/backend/declarer.dart 166:33             Declarer.test.<fn>.<fn>
  dart:async/zone.dart 1124:13                                  _rootRun
  dart:async/zone.dart 1021:19                                  _CustomZone.run
  dart:async/zone.dart 1516:10                                  _runZoned
  dart:async/zone.dart 1463:12                                  runZoned
  package:test_api/src/backend/declarer.dart 165:13             Declarer.test.<fn>
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1045:19                                  _CustomZone.registerCallback
  dart:async/zone.dart 962:22                                   _CustomZone.bindCallbackGuarded
  dart:async/timer.dart 52:45                                   new Timer
  dart:async/timer.dart 87:9                                    Timer.run
  dart:async/future.dart 174:11                                 new Future
  package:test_api/src/backend/invoker.dart 399:21              Invoker._onRun.<fn>.<fn>.<fn>

00:32 +0 -1: Meals Catalogue App (tearDownAll)
00:32 +0 -1: Meals Catalogue App Show default app bar title [E]
  DriverError: Failed to fulfill GetText due to remote error
  Original error: Bad state: The client closed with pending request "ext.flutter.driver".
  Original stack trace:
  #0      new Client.withoutJson.<anonymous closure> (package:json_rpc_2/src/client.dart:70:24)
  #1      StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
  #2      StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_
zone_specification.dart:119:48)
  #3      _rootRun (dart:async/zone.dart:1120:38)
  #4      _CustomZone.run (dart:async/zone.dart:1021:19)
  #5      _FutureListener.handleWhenComplete (dart:async/future_impl.dart:150:18)
  #6      Future._propagateToListeners.handleWhenCompleteCallback (dart:async/future_impl.dart:609:39)
  #7      Future._propagateToListeners (dart:async/future_impl.dart:665:37)
  #8      Future._propagateToListeners (dart:async/future_impl.dart:566:9)
  #9      Future._completeWithValue (dart:async/future_impl.dart:483:5)
  #10     Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:513:7)
  #11     StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
  #12     StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_
zone_specification.dart:119:48)
  #13     _rootRun (dart:async/zone.dart:1124:13)
  #14     _CustomZone.run (dart:async/zone.dart:1021:19)
  #15     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
  #16     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
  #17     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
  #18     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
  #19     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
  #20     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
  #21     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)


  package:flutter_driver/src/driver/driver.dart 424:7   FlutterDriver._sendCommand
  ===== asynchronous gap ===========================
  dart:async/future_impl.dart 22:43                     _Completer.completeError
  dart:async-patch/async_patch.dart 40:18               _AsyncAwaitCompleter.completeError
  package:flutter_driver/src/driver/driver.dart         FlutterDriver._sendCommand
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1062:19                          _CustomZone.registerBinaryCallback
  dart:async-patch/async_patch.dart 86:23               _asyncErrorWrapperHelper
  package:flutter_driver/src/driver/driver.dart         FlutterDriver._sendCommand
  package:flutter_driver/src/driver/driver.dart 560:41  FlutterDriver.getText
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1053:19                          _CustomZone.registerUnaryCallback
  dart:async-patch/async_patch.dart 77:23               _asyncThenWrapperHelper
  package:test_api/src/backend/declarer.dart            Declarer.test.<fn>.<fn>.<fn>
  package:test_api/src/backend/invoker.dart 250:15      Invoker.waitForOutstandingCallbacks.<fn>
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1045:19                          _CustomZone.registerCallback
  dart:async/zone.dart 962:22                           _CustomZone.bindCallbackGuarded
  dart:async/timer.dart 52:45                           new Timer
  dart:async/timer.dart 87:9                            Timer.run
  dart:async/future.dart 174:11                         new Future
  package:test_api/src/backend/invoker.dart 399:21      Invoker._onRun.<fn>.<fn>.<fn>

00:32 +0 -1: Some tests failed.

Unhandled exception:
Dummy exception to set exit code.
#0      _rootHandleUncaughtError.<anonymous closure> (dart:async/zone.dart:1112:29)
#1      _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#2      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#3      _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
#4      _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)
Stopping application instance.
Driver tests failed: 255
```
**flutter doctor:**
```
D:\AndroidStudioProjects\MFDEProject\meals_catalogue>flutter doctor -v
[√] Flutter (Channel master, v1.5.4-pre.17, on Microsoft Windows [Version 10.0.17134.765], locale
    en-ID)
    • Flutter version 1.5.4-pre.17 at D:\Android\Sdk\flutter_windows_v1.2.1-stable\flutter
    • Framework revision 0545c63b9b (7 weeks ago), 2019-04-19 20:41:32 -0400
    • Engine revision ca31a7c57b
    • Dart version 2.3.0 (build 2.3.0-dev.0.1 cf4444b803)


[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at D:\Android\Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-Q, build-tools 28.0.3
    • Java binary at: D:\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.4)
    • Android Studio at D:\Android\Android Studio
    • Flutter plugin version 36.0.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[√] IntelliJ IDEA Ultimate Edition (version 2019.1)
    • IntelliJ at D:\JetBrains\IntelliJ IDEA 2019.1
    • Flutter plugin version 34.0.4
    • Dart plugin version 191.6183.88

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 9 (API 28) (emulator)

• No issues found!
```

Regarding to the problem, may I know the root cause and what should I do to resolve this. Thanks.

full project : https://github.com/hanseltertiusJCU/meals_catalogue/tree/final

**Comment by VictorsAlves** [comment]
tambem estou com esse erro!

**Comment by hanseltertiusJCU** [comment]
@VictorsAlves my solution is to remove CachedNetworkImage and use NetworkImage instead. In my case, when I interact with the screen, then the integration test worked as expected.

In my case, this might be caused caused by Flutter Driver trying to find something when the data was loaded. However, when the data loaded, the application was on the process of loading the image. This is why I had to interact with the screen then the integration test ran smoothly.

Besides, avoid using FutureBuilder for integration testing as Flutter Driver cannot find the specific widget.

In your case, I don't really know about the specific problem but make sure that you are getting through the right steps.

TLDR: just avoid using FutureBuilder and CachedNetworkImage for Integration test, as well as make sure that steps to do integration testing is correct.

I hope it helps.

**Comment by jayaPrakashReddy9** [comment]
TimeoutException after 0:00:30.000000: Test timed out after 30 seconds.
For the above Exception, it might take too much time  to find the key, so try to increase the Duration time in timeout .

**Comment by muarifer** [comment]
> > TLDR: just avoid using FutureBuilder and CachedNetworkImage for Integration test, as well as make sure that steps to do integration testing is correct.

FutureBuilder isn't the main problem. When you use `Compute` or any spawn function, it's breaks the flutter_driver. Please see #24703

**Comment by mikhail-tokarev** [comment]
Hi, @hanseltertiusJCU 
are you still experiencing this issue with the latest version of Flutter?
If so please reply attaching your `flutter doctor -v`

**Comment by narendhb** [comment]
i got same error 

> Unhandled exception:
Dummy exception to set exit code.
#0      _rootHandleUncaughtError.<anonymous closure> (dart:async/zone.dart:1112:29)
#1      _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#2      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#3      _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
#4      _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
Stopping application instance.
Driver tests failed: 255

flutter doctor -v result

> [√] Flutter (Channel unknown, v1.7.8+hotfix.3, on Microsoft Windows [Version 10.0.18362.418], locale
    en-US)
    • Flutter version 1.7.8+hotfix.3 at C:\flutter
    • Framework revision b712a172f9 (4 months ago), 2019-07-09 13:14:38 -0700
    • Engine revision 54ad777fd2
    • Dart version 2.4.0


[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\ROG\AppData\Local\Android\Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = C:\Users\ROG\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[!] Visual Studio - develop for Windows (Visual Studio Community 2017 15.9.7)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2017\Community
    • Visual Studio Community 2017 version 15.9.28307.423
    X Visual Studio is missing necessary components. Please re-run the Visual Studio installer for
      the "Desktop development with C++" workload, and include these components:
        MSBuild
        VC++ 2017 version 15.9 v14.16 latest v141 tools
        Windows 10 SDK (10.0.17763.0)

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code, 64-bit edition (version 1.29.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 2.21.1

[√] Connected device (3 available)
    • Samsung Galaxy S9 • 192.168.126.101:5555 • android-x86    • Android 9 (API 28)
    • Windows           • Windows              • windows-x64    • Microsoft Windows [Version
      10.0.18362.418]
    • web               • web                  • web-javascript • Google Chrome 77.0.3865.120

**Comment by iapicca** [comment]
Hi @narendhb 
did the proposed [solution ](https://github.com/flutter/flutter/issues/33921#issuecomment-518080655)worked for you?

**Comment by iapicca** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. We are therefore reluctantly going to close this bug for now. Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  

Thanks for your contribution.Could everyone who still has this problem please file a new issue with the exact descriptions what happens, logs and the output of ' flutter doctor -v'  please.
All system setups can be slightly different so its always better to open new issues and reference related issues.

**Comment by susovanascon** [comment]
why its showing this and how to resolve it 

![image](https://user-images.githubusercontent.com/79451585/122162340-0ab2c580-ce91-11eb-9c43-27141ab9fcbc.png)
